contracts/MigrationContractStaging.cdc

```diff
@@ -0,0 +1,326 @@
+    /// status == true - insert | staged == false - replace | staged == nil - remove
+    /// NOTE: Does not guarantee that the contract code is valid Cadence
+    access(all) event StagingStatusUpdated(
+        capsuleUUID: UInt64,
+        address: Address,
+        codeHash: [UInt8],
+        contract: String,
+        status: Bool?
+    )
+
+    /********************
+        Public Methods
+     ********************/
+
+    /* --- Staging Methods --- */
+
+    /// 1 - Create a host and save it in your contract-hosting account at MigrationContractStaging.HostStoragePath
+    ///
+    /// Creates a Host serving as identification for the contract account. Reference to this resource identifies the
+    /// calling address so it must be saved in account storage before being used to stage a contract update.
+    ///
+    access(all) fun createHost(): @Host {
+        return <- create Host()
+    }
+
+    /// 2 - Call stageContract() with the host reference and contract name and contract code you wish to stage
+    ///
+    /// Stages the contract code for the given contract name at the host address. If the contract is already staged,
+    /// the code will be replaced.
+    ///
+    access(all) fun stageContract(host: &Host, name: String, code: String) {
+        let capsulePath: StoragePath = self.deriveCapsuleStoragePath(contractAddress: host.address(), contractName: name)
+        if self.stagedContracts[host.address()] == nil {
+            // First time we're seeing contracts from this address - insert the address and contract name
+            self.stagedContracts.insert(key: host.address(), [name])
+            // Create a new Capsule to store the staged code
+            let capsule: @Capsule <- self.createCapsule(host: host, name: name, code: code)
+            self.account.save(<-capsule, to: capsulePath)
+        } else {
+            // We've seen contracts from this host address before
+            if let contractIndex: Int = self.stagedContracts[host.address()]!.firstIndex(of: name) {
+                // The contract is already staged - replace the code
+                let capsule: &Capsule = self.account.borrow<&Capsule>(from: capsulePath)
+                    ?? panic("Could not borrow existing Capsule from storage for staged contract")
+                capsule.replaceCode(code: code)
+            } else {
+                // First time staging this contrac - add the contract name to the list of contracts staged for host
+                self.stagedContracts[host.address()]!.append(name)
+                self.account.save(<-self.createCapsule(host: host, name: name, code: code), to: capsulePath)
+            }
+        }
+    }
+
+    /// Removes the staged contract code from the staging environment.
+    ///
+    access(all) fun unstageContract(host: &Host, name: String) {
+        pre {
+            self.isStaged(address: host.address(), name: name): "Contract is not staged"
```

Comment:
Might be better to just do a no-op if this is the case

Reply:
Right, that makes sense